src/realm/sync/noinst/client_reset.cpp

```diff
@@ -274,14 +390,130 @@ void copy_dictionary(const Obj& src_obj, Obj& dst_obj, InterRealmValueConverter&
+        InterRealmValueConverter::ConversionResult converted_src;
+        if (convert.cmp_src_to_dst(src_obj.get_any(convert.source_col()), dst_obj.get_any(convert.dest_col()),
+                                   &converted_src, update_out)) {
+            if (converted_src.requires_new_embedded_object) {
+                Obj new_embedded = dst_obj.create_and_set_linked_object(convert.dest_col());
+                convert.track_new_embedded(converted_src.src_embedded_to_check, new_embedded);
+            }
+            else {
+                dst_obj.set_any(convert.dest_col(), converted_src.converted_value);
+            }
+        }
+    }
+}
+
+struct InterRealmObjectConverter {
+    InterRealmObjectConverter(ConstTableRef table_src, TableRef table_dst, EmbeddedObjectConverter& embedded_tracker)
+        : m_embedded_tracker(embedded_tracker)
+    {
+        populate_columns_from_table(table_src, table_dst);
+    }
+
+    void copy(const Obj& src, Obj& dst, bool* update_out)
+    {
+        for (auto& column : m_columns_cache) {
+            copy_value(src, dst, column, update_out);
+        }
+    }
+
+private:
+    void populate_columns_from_table(ConstTableRef table_src, ConstTableRef table_dst)
+    {
+        m_columns_cache.clear();
+        m_columns_cache.reserve(table_src->get_column_count());
+        ColKey pk_col = table_src->get_primary_key_column();
+        for (ColKey col_key_src : table_src->get_column_keys()) {
+            if (col_key_src == pk_col)
+                continue;
+            StringData col_name = table_src->get_column_name(col_key_src);
+            ColKey col_key_dst = table_dst->get_column_key(col_name);
+            REALM_ASSERT(col_key_dst);
+            m_columns_cache.emplace_back(
+                InterRealmValueConverter(table_src, col_key_src, table_dst, col_key_dst, m_embedded_tracker));
+        }
+    }
+
+    EmbeddedObjectConverter& m_embedded_tracker;
+    std::vector<InterRealmValueConverter> m_columns_cache;
+};
+
+void EmbeddedObjectConverter::track_if_needed(Obj e_src, Obj e_dst)
+{
+    auto get_tracked_objects_for_dst_table = [this](TableKey dst_table_key) {
+        auto it = std::find_if(embedded_verified.begin(), embedded_verified.end(),
+                               [&dst_table_key](const auto& pair) -> bool {
+                                   return pair.first == dst_table_key;
+                               });
+
+        if (it != embedded_verified.end()) {
+            return it;
+        }
+        embedded_verified.push_back({dst_table_key, {}});
+        return embedded_verified.end() - 1;
+    };
+
+    auto it = get_tracked_objects_for_dst_table(e_dst.get_table()->get_key());
+    if (it->second.find(e_dst.get_key()) == it->second.end()) {
+        embedded_pending.push_back({e_src, e_dst});
+    }
+}
+
+void EmbeddedObjectConverter::process_pending()
+{
+    std::vector<std::pair<TableKey, InterRealmObjectConverter>> converters;
```

Comment:
Why not use a std::map here?

Reply:
Conceptually, this is a map yes. I used a vector because I have observed that linear search is faster for small N. I added a comment about this.

Reply:
I don't think the difference in performance is noticeable and using `std::map<TableKey, InterRealmObjectConverter>` would make the code a bit simpler. Maybe the map type you are thinking of is `std::unordered_map`, which should be avoided. `std::map` does not use hashing. Anyway - not a showstopper